File: technical/queue/golang/queue_test.go
package queue

import "testing"

func TestEmpty(t *testing.T) {
	q := &Queue{}
	if !q.Empty() {
		t.Fail()
	}
	_, front := q.PopFront()
	if front == nil {
		t.Fail()
	}
	_, back := q.PopBack()
	if back == nil {
		t.Fail()
	}
}

func TestFronts(t *testing.T) {
	q := &Queue{}
	if !q.Empty() {
		t.Fail()
	}
	q.PushFront(0)
	q.PushFront(1)
	q.PushFront(2)
	q.PushFront(3)
	three, _ := q.PopFront()
	if v, ok := three.(int); !ok || v != 3 {
		t.Fail()
	}
	two, _ := q.PopFront()
	if v, ok := two.(int); !ok || v != 2 {
		t.Fail()
	}
	one, _ := q.PopFront()
	if v, ok := one.(int); !ok || v != 1 {
		t.Fail()
	}
	zero, _ := q.PopFront()
	if v, ok := zero.(int); !ok || v != 0 {
		t.Fail()
	}
	if !q.Empty() {
		t.Fail()
	}
}

func TestBacks(t *testing.T) {
	q := &Queue{}
	if !q.Empty() {
		t.Fail()
	}
	q.PushBack(0)
	q.PushBack(1)
	q.PushBack(2)
	q.PushBack(3)
	three, _ := q.PopBack()
	if v, ok := three.(int); !ok || v != 3 {
		t.Fail()
	}
	two, _ := q.PopBack()
	if v, ok := two.(int); !ok || v != 2 {
		t.Fail()
	}
	one, _ := q.PopBack()
	if v, ok := one.(int); !ok || v != 1 {
		t.Fail()
	}
	zero, _ := q.PopBack()
	if v, ok := zero.(int); !ok || v != 0 {
		t.Fail()
	}
	if !q.Empty() {
		t.Fail()
	}
}

func TestFIFOFront(t *testing.T) {
	q := &Queue{}
	if !q.Empty() {
		t.Fail()
	}
	q.PushBack(0)
	q.PushBack(1)
	q.PushBack(2)
	q.PushBack(3)
	three, _ := q.PopFront()
	if v, ok := three.(int); !ok || v != 0 {
		t.Fail()
	}
	two, _ := q.PopFront()
	if v, ok := two.(int); !ok || v != 1 {
		t.Fail()
	}
	one, _ := q.PopFront()
	if v, ok := one.(int); !ok || v != 2 {
		t.Fail()
	}
	zero, _ := q.PopFront()
	if v, ok := zero.(int); !ok || v != 3 {
		t.Fail()
	}
	if !q.Empty() {
		t.Fail()
	}
}

func TestFIFOBack(t *testing.T) {
	q := &Queue{}
	if !q.Empty() {
		t.Fail()
	}
	q.PushFront(0)
	q.PushFront(1)
	q.PushFront(2)
	q.PushFront(3)
	three, _ := q.PopBack()
	if v, ok := three.(int); !ok || v != 0 {
		t.Fail()
	}
	two, _ := q.PopBack()
	if v, ok := two.(int); !ok || v != 1 {
		t.Fail()
	}
	one, _ := q.PopBack()
	if v, ok := one.(int); !ok || v != 2 {
		t.Fail()
	}
	zero, _ := q.PopBack()
	if v, ok := zero.(int); !ok || v != 3 {
		t.Fail()
	}
	if !q.Empty() {
		t.Fail()
	}
}

File: technical/queue/c++/queue_test.cpp
#include <stdio.h>
#include <assert.h>
#include "queue.h"

void testEmpty();
void testFront();
void testBack();
void testFIFOFront();
void testFIFOBack();

int main() {
  testEmpty();
  testFront();
  testBack();
  testFIFOFront();
  testFIFOBack();
  printf("PASS\n");
  return 0;
}

void testEmpty() {
  queue<int> q;
  assert(q.empty());
  try {
    q.pop_front();
    assert(0);
  } catch (...) {}
  try {
    q.pop_back();
    assert(0);
  } catch (...) {}
}

void testFront() {
  queue<int> q;
  q.push_front(0);
  q.push_front(1);
  q.push_front(2);
  q.push_front(3);

  assert(!q.empty());
  assert(q.pop_front() == 3);
  assert(q.pop_front() == 2);
  assert(q.pop_front() == 1);
  assert(q.pop_front() == 0);

  assert(q.empty());
}

void testBack() {
  queue<int> q;
  q.push_back(0);
  q.push_back(1);
  q.push_back(2);
  q.push_back(3);

  assert(!q.empty());
  assert(q.pop_back() == 3);
  assert(q.pop_back() == 2);
  assert(q.pop_back() == 1);
  assert(q.pop_back() == 0);

  assert(q.empty());
}

void testFIFOFront() {
  queue<int> q;
  q.push_front(0);
  q.push_front(1);
  q.push_front(2);
  q.push_front(3);

  assert(!q.empty());
  assert(q.pop_back() == 0);
  assert(q.pop_back() == 1);
  assert(q.pop_back() == 2);
  assert(q.pop_back() == 3);

  assert(q.empty());
}

void testFIFOBack() {
  queue<int> q;
  q.push_back(0);
  q.push_back(1);
  q.push_back(2);
  q.push_back(3);

  assert(!q.empty());
  assert(q.pop_front() == 0);
  assert(q.pop_front() == 1);
  assert(q.pop_front() == 2);
  assert(q.pop_front() == 3);

  assert(q.empty());
}
